ceph_iscsi_config/common.py

```diff
@@ -154,6 +154,59 @@ def _get_ceph_config(self):
 
         return cfg_dict
 
+    def _upgrade_config(self):
+        if self.config['version'] >= Config.seed_config['version']:
+            return
+
+        if self.config['version'] <= 2:
+            self.add_item("groups", element_name=None, initial_value={})
+            self.update_item("version", element_name=None, element_value=3)
+
+        if self.config['version'] == 3:
+            iqn = self.config['gateways']['iqn']
+            gateways = {}
+            portals = {}
+            for host, gateway_v3 in self.config['gateways'].items():
+                if isinstance(gateway_v3, dict):
+                    portal = gateway_v3
+                    portal.pop('iqn')
+                    active_luns = portal.pop('active_luns')
+                    updated = portal.pop('updated', None)
+                    created = portal.pop('created', None)
+                    gateway = {
+                        'active_luns': active_luns
+                    }
+                    if created:
+                        gateway['created'] = created
+                    if updated:
+                        gateway['updated'] = updated
+                    gateways[host] = gateway
+                    portals[host] = portal
+            for _, client in self.config['clients'].items():
+                client.pop('created', None)
+                client.pop('updated', None)
+            for _, group in self.config['groups'].items():
+                group.pop('created', None)
+                group.pop('updated', None)
+            target = {
+                'disks': list(self.config['disks'].keys()),
+                'clients': self.config['clients'],
+                'portals': portals,
+                'groups': self.config['groups'],
+                'controls': self.config['controls'],
+                'ip_list': self.config['gateways']['ip_list']
```

Comment:
I think the naming of ip_list and portals is off which is not your fault but we can clean it up here. In the iSCSI spec a target side portal is the IP address and port that can be used for iscsi traffic (we will ignore port since we always use the default 3260).

In functions like in ceph_iscsi_config/gateway.py and define_gateways  we use ip_list to create TPGs which are used for iscsi traffic, so matches the iscsi spec definition, but in functions like requires_restricted_auth we use it like a gateway IP list.

In the new config definition 'portals' looks like the old gateway but we seem to use the info in it both as a gateway definition and also a iSCSI portal definition.

Are you trying to prepare this for multiple active optimized portals per gateway?

Reply:
@mikechristie No, I'm not preparing multiple active optimized portals per gateway, yet.

I'm willing to improve this, but I'll need more help / input here.

This is what we have at the moment:
````
"targets": {
        "iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw": {
            "clients": {},
            "controls": {},
            "created": "2018/12/20 13:47:49",
            "disks": [],
            "groups": {},
            "ip_list": [
                "192.168.100.201",
                "192.168.100.202"
            ],
            "portals": {
                "node1": {
                    "gateway_ip_list": [
                        "192.168.100.201",
                        "192.168.100.202"
                    ],
                    "inactive_portal_ips": [
                        "192.168.100.202"
                    ],
                    "portal_ip_address": "192.168.100.201",
                    "tpgs": 2
                },
                "node2": {
                    "gateway_ip_list": [
                        "192.168.100.201",
                        "192.168.100.202"
                    ],
                    "inactive_portal_ips": [
                        "192.168.100.201"
                    ],
                    "portal_ip_address": "192.168.100.202",
                    "tpgs": 2
                }
            },
            "updated": "2018/12/20 13:49:05"
        }
    },
````

What would you like to see renamed / moved in the config file?

Maybe we should rename `targets/<iqn>/portals` to `targets/<iqn>/gateways`, and `targets/<iqn>/ip_list` to `targets/<iqn>/portal_ip_lis`, and end up with something like:

````
"targets": {
        "iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw": {
            "clients": {},
            "controls": {},
            "created": "2018/12/20 13:47:49",
            "disks": [],
            "groups": {},
            "portal_ip_list": [
                "192.168.100.201",
                "192.168.100.202"
            ],
            "gateways": {
                "node1": {
                    "gateway_ip_list": [
                        "192.168.100.201",
                        "192.168.100.202"
                    ],
                    "inactive_portal_ips": [
                        "192.168.100.202"
                    ],
                    "portal_ip_address": "192.168.100.201",
                    "tpgs": 2
                },
                "node2": {
                    "gateway_ip_list": [
                        "192.168.100.201",
                        "192.168.100.202"
                    ],
                    "inactive_portal_ips": [
                        "192.168.100.201"
                    ],
                    "portal_ip_address": "192.168.100.202",
                    "tpgs": 2
                }
            },
            "updated": "2018/12/20 13:49:05"
        }
    },
````

Is this enought? Thoughs?

With regards to the `requires_restricted_auth` function, maybe we should use `targets/<iqn>/gateways/<host>/gateway_ip_list` instead of `targets/<iqn>/portal_ip_list`.

Does this changes address your comment?

Reply:
@mikechristie  Looks like we are missing something to make the separation between `portals` and `gateways`...

The only command that allows me to add an IP address is the following:

```
cd iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw/gateways
create node1 192.168.100.201
```

And the intention of this command is to create a `portal`, not a `gateway` right?


Maybe we should have a dedicated command do create `gateways` in the global configuraton, something like:

```
cd /gateways
create node1 <some_ip>
```

and only after creating the gateway ( e.g. "node1"), we should be able to create the corresponding portals inside the target:

```
cd iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw/portals
create node1 <same_or_other_ip>
```

Does this makes sense? Does it solve the problem?

Reply:
I think the original intent was to simplify the admin actions (consolidate the addition of a gateway and portal into a single step since it probably covers the vast majority of use-cases). Perhaps we can just add new sub-commands to the gateway node to add/remove additional portal address (i.e. first one is implicit)?

Reply:
@mikechristie  @dillaman Imagine a scenario where we want to create a gateway with a public IP, that should be used to **access the REST API**, but should **NOT** be used for **iSCSI traffic**.

I think that the only way to set this up is by creating an explicit separation between gateways and portals.

For _backward compatibility_, we can automatically create the gateway if it doesn't exist.

What about something like this?

```
/> ls /
...
- gateways
- iscsi-targets
  - iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw
    - portals
...

/> cd /iscsi-targets/iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw/portals

/iscsi-targets/iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw/portals> create node1 192.168.100.201

/iscsi-targets/iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw/portals> ls /
...
- gateways
  - node1 (192.168.100.201)
- iscsi-targets
  - iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw
    - portals
      - node1 (192.168.100.201)
...

/iscsi-targets/iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw/portals> create node1 192.168.101.1

/iscsi-targets/iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw/portals> ls /
...
- gateways
  - node1 (192.168.100.201)
- iscsi-targets
  - iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw
    - portals
      - node1 (192.168.100.201, 192.168.101.1)
...

/> cd /gateways

/gateways> create node2 192.168.100.202

/gateways> ls /
...
- gateways
  - node1 (192.168.100.201)
  - node2 (192.168.100.202)
- iscsi-targets
  - iqn.2003-01.com.redhat.iscsi-gw:iscsi-igw
    - portals
      - node1 (192.168.100.201, 192.168.101.1)
...
```